### PR TITLE
C# 11: Support for `file` scoped types.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we make support for `file` scoped types. These types are only *visible* in the file where they are declared.
```csharp
file class C1 { }
```

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Modifier.cs [hunk 1]
```diff
@@ -97,7 +100,9 @@ private static void ExtractNamedTypeModifiers(Context cx, TextWriter trapFile, I
 
         public static void ExtractModifiers(Context cx, TextWriter trapFile, IEntity key, ISymbol symbol)
         {
-            HasAccessibility(cx, trapFile, key, symbol.DeclaredAccessibility);
+            if (symbol.Kind != SymbolKind.NamedType || !((INamedTypeSymbol)symbol).IsFileLocal)
```
Comment (mbg) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Modifier.cs:103]: It's not clear to me why this new check was introduced here; perhaps a comment explaining the logic in this method would be helpful?

**Review by mbg — APPROVED** [review]
This looks reasonable to me, especially after reviewing this commit-by-commit. Initially, trying to understand the refactoring from https://github.com/github/codeql/pull/12234/commits/8b967f97061a8bda66f6b85eb24c2891ef3d6ce8 combined with the other changes was a bit tricky. 

Just one small comment about something I couldn't understand from the context of the other changes in this PR.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Modifier.cs (hunk 1 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Modifier.cs:103]: That is a good point! I will add a comment.
When `IsFileLocal` is true it means that the type has declared scope `file`.
It turns out that the symbol pertaining to `C1` in the example below has declared accessibility *internal*, which I don't think we should extract. There are similar incorrect extractions for explicit interface implementations, which is then "fixed" in the QL library layer (which leads to unwanted complexity), but I prefer that we avoid extracting the "internal" modifier in the first place.

```csharp
file class C1 { }
```

Comment (mbg) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Modifier.cs:105]: Makes sense, and the `symbol.Kind != SymbolKind.NamedType` condition is there to check that this situation applies / for the cast to succeed. Thanks!

**Review by mbg — APPROVED** [review]

**Review by hvitved — APPROVED** [review]
LGTM

**Pull request merged**